src/Sylius/Bundle/CoreBundle/EventListener/PaymentPreCompleteListener.php

```diff
@@ -43,4 +38,21 @@ public function checkStockAvailability(ResourceControllerEvent $event): void
             }
         }
     }
+
+    private function canBeCompleted(ProductVariantInterface $variant, int $quantity): bool
+    {
+        if (!$variant->isTracked()) {
+            return true;
+        }
+
+        if ($variant->getOnHold() - $quantity < 0) {
+            return false;
+        }
+
+        if ($variant->getOnHand() - $quantity < 0) {
```

Comment:
I'm not sure, but I feel that we should only check `onHold`. However, what happens if someone orders a product variant and, before completing this order, the administrator activates tracking for this variant?

Reply:
This logic comes from the `\Sylius\Component\Core\Inventory\Operator\OrderInventoryOperator::sell` method.

```php
<?php

// ...
Assert::greaterThanEq(
                ($variant->getOnHold() - $orderItem->getQuantity()),
                0,
                sprintf(
                    'Not enough units to decrease on hold quantity from the inventory of a variant "%s".',
                    $variant->getName(),
                ),
            );

            Assert::greaterThanEq(
                ($variant->getOnHand() - $orderItem->getQuantity()),
                0,
                sprintf(
                    'Not enough units to decrease on hand quantity from the inventory of a variant "%s".',
                    $variant->getName(),
                ),
            );
// ...
```

So both values are checked, same as in the example above. If you `hold` a variant, the `onHand` value is not decreased.

```
1. There're 4 socks in stock (onHand: 4, onHold: 0)
2. You buy 2 socks with using the offline payment method (onHand: 4, onHold 2)
3. As an admin you complete the payment (onHand: 2, onHold: 0)
```

Reply:
> However, what happens if someone orders a product variant and, before completing this order, the administrator activates tracking for this variant?

I guess we have to think about it, as currently it'll throw an exception. We'll discuss it on the daily meeting, and create a task if needed.

Reply:
This onHand/onHold pre-sell check logic feels important enough to varant it's own checker.

For example, why not add new method `canBeSold`  to https://github.com/Sylius/Sylius/blob/a0ad0d1e4c61ed8ec30bbb39e36184cb910736b2/src/Sylius/Component/Core/Inventory/Operator/OrderInventoryOperator.php#L52 or make brand new `OrderInventoryChecker`

Reply:
I think it needs a bit of work.
Because the https://github.com/Sylius/Sylius/blob/7735d06395e6b12b8060e72989b82409d6463516/src/Sylius/Component/Core/Inventory/Operator/OrderInventoryOperator.php#L62 or https://github.com/Sylius/Sylius/blob/7735d06395e6b12b8060e72989b82409d6463516/src/Sylius/Component/Core/Inventory/Operator/OrderInventoryOperator.php#L62 Assertions introduce a bug into Sylius.

Let me explain:
* On a Variant, stock management is activated
* A customer places an order with this variant
* An administrator deactivates stock management on this variant
* On actions, such as payment, order cancellation, etc., we get a 500 error because the assertion is not correct.